src/dbnode/storage/shard.go

```diff
@@ -1047,59 +1050,111 @@ func (s *dbShard) writeAndIndex(
 	}, nil
 }
 
-func (s *dbShard) SeriesReadWriteRef(
+func (s *dbShard) SeriesRefResolver(
 	id ident.ID,
 	tags ident.TagIterator,
-) (SeriesReadWriteRef, error) {
+) (bootstrap.SeriesRefResolver, error) {
 	// Try retrieve existing series.
 	entry, _, err := s.tryRetrieveWritableSeries(id)
 	if err != nil {
-		return SeriesReadWriteRef{}, err
+		return nil, err
 	}
 
 	if entry != nil {
 		// The read/write ref is already incremented.
-		return SeriesReadWriteRef{
-			Series:              entry,
-			Shard:               s.shard,
-			UniqueIndex:         entry.Index,
-			ReleaseReadWriteRef: entry,
-		}, nil
-	}
-
-	// NB(r): Insert synchronously so caller has access to the series
-	// immediately, otherwise calls to LoadBlock(..) etc on the series itself
-	// may have no effect if a collision with the same series
-	// being put in the insert queue may cause a block to be loaded to a
-	// series which gets discarded.
-	// TODO(r): Probably can't insert series sync otherwise we stall a ton
-	// of writes... need a better solution for bootstrapping.
-	// This is what can cause writes to degrade during bootstrap if
-	// write lock is super contended.
-	// Having said that, now that writes are kept in a separate "bootstrap"
-	// buffer in the series itself to normal writes then merged at end of
-	// bootstrap it somewhat mitigates some lock contention since the shard
-	// lock is still contended but at least series writes due to commit log
-	// bootstrapping do not interrupt normal writes waiting for ability
-	// to write to an individual series.
-	entry, err = s.insertSeriesSync(id, newTagsIterArg(tags), insertSyncOptions{
-		insertType: insertSyncIncReaderWriterCount,
-		// NB(bodu): We transparently index in the series ref when
-		// bootstrapping now instead of when grabbing a ref.
-		hasPendingIndex: false,
+		return entry, nil
+	}
+
+	entry, err = s.newShardEntry(id, newTagsIterArg(tags))
+	if err != nil {
+		return nil, err
+	}
+	entry.IncrementReaderWriterCount()
+	wg, err := s.insertQueue.Insert(dbShardInsert{
+		entry: entry,
+		opts: dbShardInsertAsyncOptions{
+			// skipRateLimit for true since this method is used by bootstrapping
+			// and should not be rate limited.
+			skipRateLimit:            true,
+			entryRefCountIncremented: false, // should be false because otherwise entry won't be inserted.
```

Comment:
Agree that this is confusing and this looks like `entryRefCountIncremented` should be set to `true` but if we would do this, then during async write new series won't be inserted because it checks for this property and makes early return if it's `true`.  I didn't want to change the logic in actual write code so the only way to make it work was to set `entryRefCountIncremented` to `false`. 
Also, initially I was thinking that there is no need to do `entry.IncrementReaderWriterCount()` but then resolve errors started to appear because new entries got expired before some of the `resolve()` calls. Adding `entry.IncrementReaderWriterCount()` fixed this issue. 
I guess we need to increment refCount here because even `s.tryRetrieveWritableSeries(id)` does the same each time it founds an existing entry.
Not sure how the comment could be improved here. Any ideas?

Reply:
I guess the comment should not tell what the consequences would be, but explain the reason why it is `false` despite the increment few lines above (not familiar enough with the logics here to suggest the actual text).

Reply:
updated comments.